### PR TITLE
BLD, BUG: Fix detecting aarch64 on macOS

### PR DESCRIPTION
**Pull request by seiko2plus**
related to #17964

TODO:
- [x] manually test on a wide range of compilers and platforms

### CODE REVIEW

**Comment by mattip** [comment]
This is OK, but I am concerned that the no-optimizations path led to a segfault. If that is the case, perhaps the detection code should raise an error earlier instead of allowing a bad build.

**Comment by charris** [comment]
Will this need a backport?

**Comment by seiko2plus** [comment]
@mattip, 
just another bug discovered by the build log in #17964. the current patch didn't fix the issue I thought the problem is `CCompilerOpt` detecting the arch as it was `armv7` but after a deep look in the build log I found that `CCompilerOpt` actually detecting CPU arch as `x86_64` because Xcode toolchain adds `-arch x86_64` to the compiler build options.
So I gonna need to fix `CCompilerOpt::dist_info` to avoid searching on command options.

@charris, yes, I think we need a backport to `1.20`.

File: numpy/distutils/ccompiler_opt.py [hunk 1]
```diff
@@ -958,8 +950,8 @@ def __init__(self):
             but still has the same gcc optimization flags.
             """
             self.dist_log(
-                "unable to detect compiler name via info <<\n%s\n>> "
-                "treating it as a gcc" % dist_info,
+                f"unable to detect compiler name which lead to treat it as gcc. "
```
Comment (seiko2plus) [inline numpy/distutils/ccompiler_opt.py:953]:
```suggestion
                "unable to detect compiler type which leads to treating it as GCC. "
                "this is a normal behavior if you're using gcc-like compiler such as MinGW or IBM/XLC."
```

File: numpy/distutils/ccompiler_opt.py [hunk 2]
```diff
@@ -579,45 +579,38 @@ def dist_test(self, source, flags):
...
-                cc_info += "x86"
-            else:
-                cc_info += "x86_64"
+            platform = "x86"
         else:
-            # the last hope, too bad for cross-compiling
-            import platform
-            cc_info += platform.machine()
+            from distutils.util import get_platform
+            platform = get_platform()
```
Comment (seiko2plus) [inline numpy/distutils/ccompiler_opt.py:598]: according to [distutils.util doc](https://docs.python.org/3/distutils/apiref.html#module-distutils.util), `get_platform()` should be friendly with osx universal binaries.

File: numpy/distutils/ccompiler_opt.py [hunk 3]
```diff
@@ -579,45 +579,41 @@ def dist_test(self, source, flags):
...
     def dist_info(self):
-        """Return a string containing all environment information, required
-        by the abstract class '_CCompiler' to discovering the platform
-        environment, also used as a cache factor in order to detect
-        any changes from outside.
+        """
+        Return a tuple conatins info about (platform, compiler, extra_args),
+        required by the abstract class '_CCompiler' to discovering the
+        platform environment, also used as a cache factor in order to detect
+        any changes happend from outside.
```
Comment (h-vetinari) [inline numpy/distutils/ccompiler_opt.py:586]:
```suggestion
        Return a tuple containing info about (platform, compiler, extra_args),
        required by the abstract class '_CCompiler' for discovering the
        platform environment. This is also used as a cache factor in order
        to detect any changes happening from outside.
```

Comment (seiko2plus) [reply numpy/distutils/ccompiler_opt.py]: Thank you

**Comment by charris** [comment]
Let's get this into 1.20.0rc2. Thanks Sayed.

**Pull request merged**